### ISSUE TITLE
Isomorphic Compatibility? Required: fully immutable client side `store`.

### ISSUE THREAD

**Issue by rbose85** [davidkpiano/react-redux-form#59]
...

**Comment by davidkpiano** [comment]
Sorry, can you elaborate?

**Comment by rbose85** [comment]
:wink: Sorry. My eager fingers hit enter too quickly. I'm just jotting down my post and I'll update the ticket. Thanks!

**Comment by rbose85** [comment]
Awesome library! Love what you're doing for forms in the react + redux echo system. However, I'm having a slight issue with getting `Immutable` working in an isomorphic project.

## Prototype

Extending the code snippet you posted on [Stack Overflow](http://stackoverflow.com/a/35685318) I was successful in using the `modelReducer`, exported at `react-redux-form/lib/immutable`, to work with my immutable `user` model in the store. Within a sandboxed project I was even able to add the `userForm` object to the store ([as suggested in the docs](https://davidkpiano.gitbooks.io/react-redux-form/content/form_reducers.html)).

```js
// store.js
import { createStore, combineReducers } from 'redux'
import { formReducer } from 'react-redux-form'
import { modelReducer } from 'react-redux-form/lib/immutable'
import Immutable from 'immutable'

export default createStore(combineReducers({
  user: modelReducer('user', Immutable.Map({
    email: '',
    phone: ''
  })),
  userForm: formReducer('user')
}))
```

## Production

Moving out of the sandbox, I am attempting to integrate the above working example into my production app. The app is isomorphic, with a fully immutable client-side store. Isomorphism is implemented using something very close to the `redux` [recommended recipe](http://redux.js.org/docs/recipes/ServerRendering.html):

```js
// client.js
import React from 'react'
import { render } from 'react-dom'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import App from './containers/App'
import reducer from './reducer'
import Immutable from 'immutable'                            // <-- new line

const initialState = window.__INITIAL_STATE__

// over-the-wire JS/JSON hydrates into `store` as immutable
const immutableInitialState = Immutable.fromJS(initialState) // <-- new line

const store = createStore(reducer, immutableInitialState)

render(
  <Provider store={store}>
    <App />
  </Provider>,
  document.getElementById('root')
)
```

From the above extract `immutableInitialState` will reference an immutable object, and that is being used to hydrate the store.

## Problem

`userForm` in `store.js` will be an immutable type once the client-side store hydration has completed for my isomorphic app. When I then click into a field on my form I get the following error in the console, produced by the form's reducer:

> `Uncaught Error: Error when trying to retrieve field 'email from an invalid/empty form state. Must pass in a valid form state as the first argument.`

Having dug into the source code of `react-redux-form` I believe the issue lies in the `getField()` function of `reducers/form-reducer.js`. I believe this function is executing on the assumption that the value of the `state` parameter is a plain JavaScript object - whereas, in my case, `state` references an immutable representation of the same object due to my client-side store hydration.

## Proposal

Please review my issue and let me know what your thoughts.

If it's okay with you, I would like to fork `react-redux-form` and work on a PR that would enhance the current `getField()` function. I would like to upgrade `getField()` such that it can work with `state` when `state` is a plain JS object, and when `state` is an immutable.

Before starting such work may I ask that you confirm if updating the `getField()` function in this way is the cleanest means to solve this problem.

**Comment by davidkpiano** [comment]
That makes sense. I'd probably want the `getField()` function to be placed in this file: https://github.com/davidkpiano/react-redux-form/blob/master/src/immutable/index.js if possible?

That way, you can use it like so:

```js
import { getField } from 'react-redux-form/lib/immutable';

// ...
```

What do you think about that?

**Comment by rbose85** [comment]
hmm .. do you mean a 2nd `getField()` function in `immutable/index.js` related to getting on immutables, and have `getField()` in `reducers/form-reducer.js` call into `immutable/index.js` for the scenario that the `state` is immutable? Or would you like for just the one `getField()` function, and have it moved from `reducers/form-reducer.js` to `immutable/index.js`?

**Comment by rbose85** [comment]
Scratch that. I think I get what you mean. PR coming up ...

**Comment by davidkpiano** [comment]
Yep, separate `getField()` function, so that the core of RRF remains unopinionated. :)

**Comment by rbose85** [comment]
Okay, I know where #63 is going wrong for me ... I may need to change the structure of the logic to the `formReducer` ... Are you opposed to adding a project dependency on `Immutable`? I appreciate you're wanting to make `react-redux-form` a modular project.

**Comment by davidkpiano** [comment]
I would rather avoid having to make it depend on `Immutable`. Would you want to instead have a decorator function that converts the form state `.toJS`, and then pass it to the original `formReducer` function, and convert it to an immutable data structure back out again?

**Comment by rbose85** [comment]
@davidkpiano, thanks for all your help on this. I've decided to simplify my needs for now, and go in a different direction.

I did have a go at the refactor on #63. However, the more I looked at it the more I realised it needs a big refactor - to the point where wasn't sure of the gain .. I too thought it might be as simple a case of `.toJS()`, but it's not. It's the logic of the `formReducer` itself. Because the state of the form and the state of the fields of the form are updated separately. :(

Thanks again. I'm happy to close this.
